Ska3 and chuck the MS stuff

jeanconn opened the pull request:
Update to Py3 and fundamentally reorganize.

This attitude_error_mon was designed to make a plot of one shots and then look at attitude errors over dwells before and after the transition to ignoring the multiple star suspected flag.  The multiple star suspected flag is moot at this point, and the SSAWG request (which was still being refined) was to look at perhaps a year of data before the last IRU calibration update and compare that with the attitude errors since the last calibration update.

This new code:
  - cuts all pieces related to multiple star suspected flag
  - reorganizes processing of dwell data and moves more strictly to using kadi maneuvers, as we no longer care if the dwell before a one shot was respecting the MS flag
  - adds new code to put outliers in a table on the html page with comments
       - code borrowed from kalman_watch 
      - outliers currently flagged if 99th percentile point_err > 20 arcsec or roll_err > 40 arcsec
  - leaves the autoscale for the point and roll error plots t in place with the idea that it should be easy to come back and tweak the limits or the outlier collection if needed

Do we want
  - [x] one shot plot with pitch and yaw [ no ] 
  - [x] plot with aberration correction [ yes but in new PR ]

Testing:

  - [x] Run over the last year + 60 days of data and with the 2020:129 IRU cal as a reference. 
      - [x] Makes reasonable plots and outlier table
   - [x] Run over a longer chunk of the mission to confirm good behavior of the outlier code

jeanconn commented:
This is probably ready for feedback... specifically if you want me to go much further with cleanup, documentation, modernization, or if this is sufficient to just get it going again.

taldcroft commented on att_err_mon.py:
```diff
@@ -1,158 +1,150 @@
+        obs['one_shot'] = m.one_shot
+        obs['one_shot_yaw'] = m.one_shot_yaw
+        obs['one_shot_pitch'] = m.one_shot_pitch
+        obs['dwell_duration'] = DateTime(m.get_next().start).secs - DateTime(m.npnt_start).secs
```
Is there a reason you can't use `m.next_nman_start`?

jeanconn replied:
I don't think so.  There are a lot of attributes in there; is that one better?

taldcroft replied:
Yes, `next_nman_start` is better since it doesn't require that there be a next entry in the database and it is more explicit. The `start` attribute refers to the actual start of the maneuver. Being fussy, what we want here is the start of NMAN for the next manuever.

The maneuver events are only put into the database if they are "complete", meaning that the "next" maneuver has been seen in telemetry and the `next_man_start` field can be filled in correctly.

jeanconn replied:
OK. Though I just had this loop breaking a few lines up if get_next() was null so they somewhat became equivalent?  Also, regarding NMAN transition, the times used for the range for the point/roll errors are more explicit anyway (and trim); this rough dwell duration is really just used for the 1000s filter (though if we want something else would be good to know anyway).

taldcroft replied:
You shouldn't need the `get_next()` test above if you don't use `get_next()` here.  More win for code simplicity.

taldcroft commented on att_err_mon.py:
```diff
@@ -161,71 +153,105 @@ def att_err_time_plots(ref_data, msd_data, min_dwell_time=1000, outdir='.'):
         if ax == 'roll':
             bin_width = .25
-            lim = np.max([12, np.max(msd_data['roll_err'])])
+            lim = np.max([12, np.max(recent_data['roll_err'])])
```
Does there need to be a mechanism in place to prevent this auto-scaling from reacting to one bad (large) data point? In the previous version the plots just became useless for this reason.

One idea might be to keep a count of points that are "too big", clip the points within range, and if count_big > 0 then include some text.

jeanconn replied:
Indeed, I thought I somewhat had that by the "outlier" mechanism in the calling method.

The problem with this monitor in general was, I thought, that we aren't sure if we are interested in changes in the distribution on a small scale, or increase in frequency of outliers, or some other behavior.  With that in mind, I'm not sure about the best plot limits.  But the calling code borrowed from kalman_watch to make a list of outliers for review.

taldcroft replied:
OK, I didn't notice that about the outliers, I just ran the code, saw that it worked and looked at the output page. (Would be good to get more formal with your PR descriptions and call out new things like that at the top level possibly in an itemized list, and then include your testing as a separate section).

Have you tested the outliers code?

I think the way this current code is written you will still autoscale to the size of the worst outlier, thus getting potentially back into the situation of useless plots.  But I also don't know if you have addressed the processing problems that allowed for bad data points to get into the plotted dataset.

jeanconn replied:
The outliers code is working nicely to pick up 47190.  I assume it showed up on your output page?  

I tried to pick some kind of reasonable compromise on the limits for the outliers, but the problem with this kind of code is that we still don't know what weird things will be interesting. And yes, it will still autoscale if we have large values that are not otherwise anomalous.  I don't blame the code that much for previously including MUPS test fires (I don't think I can filter those with kadi).  With a rough understanding that *maybe* those (and our recent huge attitude error going to safe mode) are the only kinds of bad data, the new outlier printing might be enough.  

I also considered using the method you have in the aimpoint code by which you put placeholder points for points that have left the plot.  In the end I thought we didn't care enough for that complexity.

jeanconn commented:
Does your 60 days comment mean you disagree with the "use a supplied date" way that I've set up the code with the intent of using the last IRU CAL update as the reference time per the SSAWG discussion?  The discussion went both ways IIRC.

jeanconn commented on att_err_mon.py:
```diff
@@ -1,158 +1,150 @@
             err.remove_intervals(events.dumps)
             events.tsc_moves.interval_pad = (0, 300)
             err.remove_intervals(events.tsc_moves)
+            err.remove_intervals(events.dark_cal_replicas)
```
I should cut this; I think the attitude errors aren't updated during the NMAN intervals of the SRDCs anyway, so the filter here doesn't buy us anything.  Do we want to try to trim *around* those intervals?  I don't think we care.

taldcroft replied:
👍

jeanconn commented on index_template.html:
```diff
@@ -10,7 +10,7 @@
 <h2>One shots and attitude errors</h2
 
-<h3>(Updated 
+<h3>(Updated
```
And I meant to just write out the processing time from the jinja2 but the flastmod seems fine.  Need to put back the space.

jeanconn commented:
Also, for the default of "60 days" should it compare the last 60 days to the year before those 60 days started?  Or include the 60 days in the year?

taldcroft commented:
> Also, for the default of "60 days" should it compare the last 60 days to the year before those 60 days started? Or include the 60 days in the year?

Keep it simple.  Just compare the last 60 days to the last 365 days.

taldcroft commented:
The standard trending plot we look at in SSAWG is intended to highlight potential recent performance changes in the hardware.  It is not to evaluate the efficacy of an IRU cal patch. So the default plot for trending should be the 60 / 365 plots.  FOT PCAD does the (much-more-detailed) evaluation of IRU performance broken in intervals corresponding to IRU cal "epochs", so this page doesn't need to replicate that.

taldcroft commented:
We do not need the one-shot plot broken down by pitch and yaw, but adding the aberration-corrected version is a good idea to take out noise and give a more accurate view of gyro performance.

taldcroft commented:
> leaves the autoscale for the point and roll error plots t in place with the idea that it should be easy to come back and tweak the limits or the outlier collection if needed

This should be changed to just use fixed limits for the plots, presumably the 20 and 40 arcsec. The outlier table eliminates the need for autoscaling, though it might make sense to *clip* data points so they show up in the plot, understanding that points at the limit correspond to those in the table.

We have good historical data to show that problems with this plot autoscaling might not get fixed in a timely manner due to other priorities.

jeanconn commented:
I disagree a bit in that I think there can be interesting data with changes in aggregate behavior much below the 20 and 40 arcsec limit... but not sure if a 5 arcsec point on the plot, for example, deserves to make it into the outlier list.  I'm still trying not to beg the questsion.  I think the plot autoscaling issues were all due to MUPS test fires that were large outliers, so the historical data suggests that issue has been fixed with the current code...?

jeanconn commented:
Also with regard to historical data and fixing in a timely manner, I mostly didn't take the 1 minute to fix anything here because we'd decided we didn't care about the plots at all.

taldcroft commented:
- The current 7.5 and 15 arcsec plot and outlier limits are good.
- I think the data storage / caching needs some work because the current scheme seems to always append new data to the current file. That is predicated on a fixed reference date, not a rolling date like 60 days ago. If I run the program once with `--recent-start=-60` and then run again with `--recent-start=-30`, then two plots are exactly the same.
- One idea is to drop the separate recent data file and only have a single file of data points for the last year. Then only in the plotting functions is there is distinction between "long-term" and "recent".
- Looking at the plots now, I'm going to reverse myself and say that the two data sets (1 year and 60 days) should be distinct.  So it would be -1 year to -60 days, and -60 days to present. This will still be simple enough if the "recent" data file is dropped.
- I don't think there is value-added in the separate time series plot of the roll error and point error magnitudes for the last year vs. last 60 days.  Basically you can see the last 60 days clearly in the 1-year plot.
- Can you update the section title from `Attitude errors for nominal dwells greater than 1ks.` to `Attitude errors (AOATTER 1,2,3)  for nominal dwells greater than 1ks.`  The word "attitude errors" is fairly ambiguous.  I was surprised at how small they are (esp. roll), but AOATTER is just the diff between commanded and estimated (I think!), which is different and smaller than the true attitude error.
- For this tool the IRU cals are not what I want to see. We can revisit with Eric if needed, but note that the IRU cal is completely unrelated to the second half of this page (the roll / point errors), so there is already a mismatch. The baseline need for this tool is to see trends in the IRU. If you stick with a fixed-date cutoff of "recent" being since the last IRU cal update, then that could be 2 years old or 5 days old, depending on when the last cal was uplinked. This is not helpful to the goal here.
- What does "This is not filtered" mean? 
- When I run this locally I get about 30 "outliers" with values of 999.0.  Is that expected?
- Ideally the safe mode data should not be included at all. You can use `kadi.events.safe_suns` and exclude everything from the `start` to the `stop` plus a day (or so).

jeanconn commented:
I think this wasn't ready for re-review.

jeanconn commented:
Regarding "Ideally the safe mode data should not be included at all. You can use kadi.events.safe_suns and exclude everything from the start to the stop plus a day (or so)."  I looked at that, but in the end figured that we might actually be interested in time before or after the event, so picking some arbitrary padding seemed less fun than just keeping it and annotating it.

jeanconn commented:
Re "What does "This is not filtered" mean?" means that the one shots are not filtered by dwell duration.  But that's a holdover from the previous code that can be made more explicit.

jeanconn commented:
Re: "I don't think there is value-added in the separate time series plot of the roll error and point error magnitudes for the last year vs. last 60 days. Basically you can see the last 60 days clearly in the 1-year plot."  Do you just mean you don't see a reason to break them out on separate axes with separate colors?  The whole point was to highlight comparison.  If we want to chuck that ... that's fine.

jeanconn commented:
Update text on the page to be less terse.

jeanconn commented:
I think I've updated per the block of comments from @taldcroft above.  I have one action item:

  - [ ] review AOATTER data and see if any actually-bogus data is in the computation for the obsid at safe mode transition.  I would suggest that any bogus or wrong data eng_archive data should probably be in ltt_bads anyway.  Or have I misunderstood the point of ltt_bads?

I updated some of the figure titles to be a bit more consistent with the page text, but did not add a narrative section (per the "update page to be less terse" directive), as I just didn't have good ideas about what to say that would be relevant at this point.

taldcroft commented:
LTT_bads is one of those ideas that seemed like a good idea at the time but there wasn't a plan for maintaining it.  The last LTT bad interval was in 2015, so basically you shouldn't bother with that. Also it appears to be rather crude in that it eliminates one full calendar day at a time.
